### ISSUE TITLE
[Feature Request] expose 'resolveModule' as an API

### ISSUE THREAD

**Issue by hanayashiki** [evanw/esbuild#1773]
For a TS/JS interpreter using lazy loading has the following situation:

1. Read a file
2. Resolve, load and transpile its dependencies recursively.
3. Execute the file, and possibly repeat the progress again when there's async import.

And that's what https://github.com/antfu/esbuild-node-loader is trying to do.

Currently, there's repeated problem with the 'resolve' function. We are now using the original typescript resolver, but it is not always the case it works, e.g., ts resolve eagerly accepts '.d.ts' result rather than '.js', which is useful with type-checking but incorrect if we want to resolve the runnable file. Bundling doesn't work because some libraries assume unbundled environment when executed with node.

Esbuild has done a good job resolving typescript and javascript modules, according to my experience, but as far as I know, it doesn't expose the API for resolving. I wish the author of this fantastic library could take this into consideration.

Thank you.

**Comment by hyrious** [comment]
Invoking the resolver in go requires additional IPC resource and must be async, maybe rewriting [resolver.go](https://github.com/evanw/esbuild/blob/01b65ec5f61a086c4d99bd59d7d776804258b088/internal/resolver/resolver.go#L262) in js is a better choice.

You can make a simple wrapper of esbuild to write an async resolver like such:
```js
const esbuild = require('esbuild')

async function esbuildResolve(id, dir) {
  let result
  await esbuild.build({
    stdin: {
      contents: `import ${JSON.stringify(id)}`,
      resolveDir: dir,
    },
    write: false,
    bundle: true,
    plugins: [{
      name: 'resolve',
      setup({ onLoad }) {
        onLoad({ filter: /.*/ }, args => {
          result = args.path
          return { contents: '' }
        })
      }
    }]
  })
  return result
}

;(async () => {
  let a = await esbuildResolve('./a', __dirname)
  console.log(a)
})();
```

**Comment by hanayashiki** [comment]
> Invoking the resolver in go requires additional IPC resource and must be async, maybe rewriting [resolver.go](https://github.com/evanw/esbuild/blob/01b65ec5f61a086c4d99bd59d7d776804258b088/internal/resolver/resolver.go#L262) in js is a better choice.
> 
> You can make a simple wrapper of esbuild to write an async resolver like such:
> 
> ```js
> const esbuild = require('esbuild')
> 
> async function esbuildResolve(id, dir) {
>   let result
>   await esbuild.build({
>     stdin: {
>       contents: `import ${JSON.stringify(id)}`,
>       resolveDir: dir,
>     },
>     write: false,
>     bundle: true,
>     plugins: [{
>       name: 'resolve',
>       setup({ onLoad }) {
>         onLoad({ filter: /.*/ }, args => {
>           result = args.path
>           return { contents: '' }
>         })
>       }
>     }]
>   })
>   return result
> }
> 
> ;(async () => {
>   let a = await esbuildResolve('./a', __dirname)
>   console.log(a)
> })();
> ```

smart hack! thank you for that

**Comment by evanw** [comment]
Closing this as a resolve API has now been implemented: #1881. Path resolution depends on the build configuration so it's not something that makes sense to expose as a separate top-level API. It as been exposed as a plugin API instead.
